Commenting

jamesMckinnonSchool opened the pull request:
## Summery

- Commenting functionality is complete. Users can now click the comment icon on a meed event on the mood feed to view and create comments for that mood event.
- Can also navigate to comments from mood details page.
- When a mood is created, username is now added as a field for each mood event so that it is available without delay for mood event cards on home feed and for comments.
- Added a username variable to the MoodEvent class.
- Updated MoodAdapter to add the comment bubble icon, username text and click listener.
- Improved UI of mood history page. Probably should have made a new branch for this.
- Will implement tests soon.

Resolves: #217 
Resolves: #216 
Resolves: #215
Resolves: #214
Resolves: #213
Resolves: #212 

## Testing

- How should a reviewer test this feature, fix, or UI?
    - Create mood events and then make comments in them. 
    - Navigate through all of the relevant pages. Home > comments > mood details > comments button > comments. 
- Are there any specific commands, steps, or credentials needed?
    - I added an index in firebase for sorting mood comments by time.
    - I added to the rules in the DB for the new comments collection.

## Merge Instructions

- Any considerations that other PR's may need to take into note? 
    - @TahaKamil I modified the mood history page slightly but I don't think it will impact what you've done with filtering.
- Should the reviewer delete the branch on merge? Y/N
    - N. Still need to add tests but want to get this out because there are changes to MoodEvent and the DB setup which could impact other features.

---

**Reviewer's Checklist**
- [x] Are all relevant issues referenced by this PR?
- [x] Does the app still build (locally)?
- [x] Do all tests pass?
- [x] Does the feature work as expected?
- [x] Does this feature preserve the app’s existing stability?  
- [x] Is the description of the changes correct/present?
- [ ] Have new tests been created or existing tests updated as needed?

jamesMckinnonSchool committed:
Commenting implemented

Commit:
MoodHistory Filtering and Test Case

# Conflicts:
#	code/app/src/main/java/com/kernelcrew/moodapp/data/MoodEventFilter.java
#	code/app/src/main/java/com/kernelcrew/moodapp/ui/components/FilterBarFragment.java

Commit:
MoodHistory Filtering and Test Case

# Conflicts:
#	code/app/src/main/java/com/kernelcrew/moodapp/data/MoodEventFilter.java
#	code/app/src/main/java/com/kernelcrew/moodapp/ui/components/FilterBarFragment.java

Commit:
JavaDoc - FilterBarFragment

Commit:
JavaDoc - FilterBarFragment

jamesMckinnonSchool committed:
Improved UI for commenting and mood history

jamesMckinnonSchool committed:
Update android.yml

jamesMckinnonSchool committed:
Update android.yml

jamesMckinnonSchool committed:
Update android.yml

TahaKamil committed:
Filtering UI Implementation (#224)

* MoodHistory Filtering and Test Case

* Forgot that List<> -> Set<>

* Added scrollTo()

* JavaDoc - FilterBarFragment

* JavaDoc - MoodEventFilter

---------

Co-authored-by: “Taha <“tkamil@ualberta.ca”>
# Conflicts:
#	code/app/src/main/java/com/kernelcrew/moodapp/ui/MoodHistory.java
#	code/app/src/main/res/layout/fragment_mood_history.xml

jamesMckinnonSchool committed:
merged new changes from main after merging filtering branch

jamesMckinnonSchool committed:
merged commenting branch

jamesMckinnonSchool committed:
fixed conflicts with mood history fragment after filtering integration

jamesMckinnonSchool committed:
Merge branch 'main' into commenting

jamesMckinnonSchool committed:
Small UI fix in mood history.

jamesMckinnonSchool committed:
updated tests to account for new username field in MoodEvent

TahaKamil commented on code/app/src/main/java/com/kernelcrew/moodapp/data/UserProvider.java:
```diff
@@ -33,6 +35,27 @@ public static UserProvider getInstance() {
+     * @return The user's username.
+     */
+    public Task<String> getUsername(@NonNull String uid) {
+        return db.collection("users").document(uid).get()
```
This does not work! Instead do:
```java
return db.collection("usernames")
                .whereEqualTo("uid", uid)
                .get()
                .onSuccessTask(querySnapshot -> {
                    if (!querySnapshot.isEmpty()) {
                         return Tasks.forResult(querySnapshot.getDocuments().get(0).getString("username"));
                    } 
                });
```

TahaKamil reviewed (dismissed):
Instead of creating a entirely new collections called "comments" in the database, please instead store the comments as an array list within the "moodEvents". This can be done like [this](https://stackoverflow.com/questions/40055764/setting-arrays-in-firebase-using-firebase-console).

Also the comments currently do not work and that is mainly because you never updated the firebase rules, make sure to change them or review them as they are currently the only thing blocking your comments!

TahaKamil commented:
Filtering doesnt seem affected from what I tested, let me know if there are any issues that occur.

jamesMckinnonSchool committed:
Update firestore.rules

jamesMckinnonSchool committed:
fixed bug that assigned wrong uid to comment

jamesMckinnonSchool committed:
merge conflicts resolved

jamesMckinnonSchool committed:
fixed test to account for new uwername field in MoodEvent

PossiblyAShrub commented on .github/workflows/android.yml:
```diff
@@ -62,45 +51,39 @@ jobs:
           target: google_apis
           script: scripts/ci.sh ui-test
-
+          emulator-options: -no-snapshot-save -no-window -gpu swiftshader_indirect -noaudio -no-boot-anim -camera-back none
```
Why did you need to add these?

Also, why did you remove all the line spacing??

jamesMckinnonSchool replied:
Ah, I was trying to debug why the test wasn't passing and this was something I found online. I ended up figuring out what the actual issue was and forgot to change this back to the original testing file. I'll change that back now.

PossiblyAShrub commented on code/app/src/main/java/com/kernelcrew/moodapp/data/MoodEvent.java:
```diff
@@ -23,6 +23,7 @@
 public class MoodEvent implements Serializable {
     private String id;
     private String uid;
+    private String username;
```
This is redundant. We should just fetch the username via the uid.

jamesMckinnonSchool replied:
I think it makes the most sense to always have the username available for mood events instead of having to bring it in from the db every time a mood event is loaded. I was loading it from the db every time before and there was sometimes a delay where the spot where the username goes would be blank while it was loading into mood events on the home feed. 

I could probably wait until all db data was loaded before displaying mood events on home feed or something so it doesn't show blank username but this works and there's no separate load time for usernames this way.

PossiblyAShrub replied:
Okay. Another option is just to maintain a client-side cache of all the uid <-> username pairs. But, given that our scope is now fixed, I suppose it's okay to denormalize this... LGTM

PossiblyAShrub commented on code/app/src/main/java/com/kernelcrew/moodapp/ui/MoodComments.java:
```diff
@@ -0,0 +1,303 @@
+    /**
+     * Refreshes the comments list by fetching the latest comments from Firestore
+     */
+    private void refreshComments() {
```
It would be best to listen to changes like we do for `MoodEventProvider::addSnapshotListener`.

PossiblyAShrub commented on code/app/src/main/java/com/kernelcrew/moodapp/ui/MoodComments.java:
```diff
@@ -0,0 +1,303 @@
+     * Uses the NavController to pop the back stack, returning to the previous fragment.
+     */
+    private void handleBackButton() {
+        androidx.navigation.fragment.NavHostFragment.findNavController(this).popBackStack();
```
Can't we import `NavHostFragment` instead of qualifying it here?

PossiblyAShrub commented on code/app/src/main/java/com/kernelcrew/moodapp/data/CommentProvider.java:
```diff
@@ -0,0 +1,73 @@
+                .whereEqualTo("moodEventId", moodEventId)
+                .orderBy("created", Query.Direction.DESCENDING)
+                .get()
+                .continueWith(task -> {
```
It would be cleaner (and more reliable) to use `onSuccessTask`. For example, see https://github.com/cmput301-w25/project-kernelcrew/blob/7d37c5b2de16890b5807dcde23df2c382809c1be/code/app/src/main/java/com/kernelcrew/moodapp/data/MoodEventProvider.java#L91-L98

PossiblyAShrub commented on code/app/src/main/java/com/kernelcrew/moodapp/data/Comment.java:
```diff
@@ -0,0 +1,88 @@
+        this.commentText = commentText;
+    }
+
+    public String logging() {
```
Can just be `toString()`. That would make more sense. Otherwise, we should add a JavaDoc.

PossiblyAShrub commented on code/app/src/main/java/com/kernelcrew/moodapp/ui/MoodComments.java:
```diff
@@ -0,0 +1,303 @@
+    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
+        View view = inflater.inflate(R.layout.fragment_mood_comments, container, false);
+
+        provider = MoodEventProvider.getInstance();
```
We also assign `provider` an `userProvider` in `onCreate`. We should only need to do that once.

PossiblyAShrub reviewed (changes requested):
Looks good so far. Don't forget to add tests (I'm okay with that being in another PR.)

jamesMckinnonSchool committed:
Update android.yml

PossiblyAShrub committed:
Remove trigger, change reason validation in mood event form (#233)

* Remove trigger, photoURL from mood event

* Change reason validation to <=200 chars
# Conflicts:
#	code/app/src/main/java/com/kernelcrew/moodapp/data/MoodEvent.java

PossiblyAShrub committed:
Add public/private toggle to mood event form (#232)

* Fix spacing caused by empty photo button and photo error views

* Add mood event visibility to mood events

* Add mood event visibility toggle to mood event form

* Add tests

* Make test work on small phone where mood selector overflows

adityadipakpatel commented:
@jamesMckinnonSchool 
regarding the merge conflict in MoodEvent.java, i believe the latest one should have these fields(as in the code block below) because trigger, photo has now been removed, see #233 :
```
    public MoodEvent(String uid, String username, Emotion emotion, String socialSituation, String reason,
                     Double latitude, Double longitude) {
```

<img width="861" alt="image" src="https://github.com/user-attachments/assets/d536cdc2-10a6-4e31-8a65-745a0701324b" />

TahaKamil commented:
> @jamesMckinnonSchool regarding the merge conflict in MoodEvent.java, i believe the latest one should have these fields(as in the code block below) because trigger, photo has now been removed, see #233 :
> 
> ```
>     public MoodEvent(String uid, String username, Emotion emotion, String socialSituation, String reason,
>                      Double latitude, Double longitude) {
> ```
> 
> <img alt="image" width="861" src="https://private-user-images.githubusercontent.com/69451242/425722841-d536cdc2-10a6-4e31-8a65-745a0701324b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI2Mzc5MzcsIm5iZiI6MTc0MjYzNzYzNywicGF0aCI6Ii82OTQ1MTI0Mi80MjU3MjI4NDEtZDUzNmNkYzItMTBhNi00ZTMxLThhNjUtNzQ1YTA3MDEzMjRiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzIyVDEwMDAzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg5NTNjMDI4M2U5MDRmOWQ5MGY3OTU3NDZlMDcxNjU4ZjNmZjUyMDEyZGY4MmZhNDc0ZmMwNjJjZGU3MGE4NGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.g0dCktMjMXAGub64BXDDP_HcBefsBRs33jc3WGMT_7A">

Thank you, I saw this message and I was also doing the same thing...saved me from a headache.

jamesMckinnonSchool committed:
before merge main

jamesMckinnonSchool committed:
made changes based on PR comments

jamesMckinnonSchool committed:
Merge branch 'main' into commenting
